lib/stages/ringmap.cpp

```diff
@@ -0,0 +1,586 @@
+                                 + ".\n"
+                                   "One of them is a nullptr (0): prod "
+                                 + std::to_string(pstate != nullptr) + ", input "
+                                 + std::to_string(istate != nullptr) + ", stack "
+                                 + std::to_string(sstate != nullptr));
+
+    if (sstate == nullptr)
+        throw std::runtime_error("MapMaker requires visibilities stacked ");
+
+    stacks = sstate->get_stack_map();
+    prods = pstate->get_prods();
+    inputs = istate->get_inputs();
+    freqs = fstate->get_freqs();
+
+    num_stack = sstate->get_num_stack();
+}
+
+bool mapMaker::setup(size_t frame_id) {
+
+    // Wait for the input buffer to be filled with data
+    if (wait_for_full_frame(in_buf, unique_name.c_str(), frame_id) == nullptr) {
+        return false;
+    }
+
+    auto in_frame = visFrameView(in_buf, frame_id);
+    ds_id = in_frame.dataset_id;
+    change_dataset_state(ds_id);
+
+    // TODO: make these config options ?
+    num_pix = 512; // # unique NS baselines
+    num_pol = 4;
+    num_time = 24. * 3600. / (in_frame.fpga_seq_length * 2.56e-6);
+    num_bl = (num_stack + 1) / 4;
+
+    sinza = std::vector<float>(num_pix, 0.);
+    for (uint i = 0; i < num_pix; i++) {
+        sinza[i] = i * 2. / num_pix - 1. + 1. / num_pix;
+    }
+
+    // generate map making matrices
+    gen_matrices();
+
+    // initialize map containers
+    mtx.lock();
+    for (auto f : freqs) {
+        std::vector<std::vector<float>> vis(num_pol);
+        std::vector<std::vector<float>> w(num_pol);
+        for (uint p = 0; p < num_pol; p++) {
+            vis.at(p).resize(num_time * num_pix);
+            w.at(p).resize(num_time);
+            std::fill(vis.at(p).begin(), vis.at(p).end(), 0.);
+            std::fill(w.at(p).begin(), w.at(p).end(), 0.);
+        }
+        map.insert(std::pair<uint64_t, std::vector<std::vector<float>>>(f.first, vis));
+        wgt.insert(std::pair<uint64_t, std::vector<std::vector<float>>>(f.first, w));
+    }
+    mtx.unlock();
+
+    // Make sure times are empty
+    times.clear();
+    times_map.clear();
+
+    // Initialize the time indexing
+    min_ctime = ts_to_double(std::get<1>(in_frame.time));
+    max_ctime = min_ctime;
+    latest = modulo<size_t>(num_time);
+
+    return true;
+}
+
+void mapMaker::gen_matrices() {
+
+    // calculate baseline for every stacked product
+    ns_baselines.reserve(num_bl);
+    chimeFeed input_a, input_b;
+    for (size_t i = 0; i < num_bl; i++) {
+        stack_ctype s = stacks[i];
+        input_a = chimeFeed::from_input(inputs[prods[s.prod].input_a]);
+        input_b = chimeFeed::from_input(inputs[prods[s.prod].input_b]);
+        ns_baselines[i] = feed_sep * (input_b.feed_location - input_a.feed_location);
+        if (s.conjugate)
+            ns_baselines[i] *= -1;
+    }
+
+    // Construct matrix of phase weights for every baseline and pixel
+    for (auto f : freqs) {
```

Comment:
It looks like you don't do any form of weighting or apodisation of the visibilities before map making, which I think explains the large amount of NS sidelobes that can be seen in the images. We should probably start using something (potentially configurable). It's a reasonably straightforward change you just multiply the baselines axis by some kind of apodization function. A Blackmann window is probably a good choice here as we just want to present clean looking maps for diagnosis, and not analysis.

Reply:
Yeah this would be really easy to implement. I think I will play around in a notebook with some different options to better wrap my head around apodization in the beamforming.